[ML] Retry datafeed searches on skipped CCS clusters

droberts195 opened the pull request:
If a remote cluster has the skip_unavailable option set and is
unavailable during a datafeed search then we should retry rather
than risk using incomplete search results. Using incomplete search
results could lead to spurious anomalies being found that do not
make sense once the remote cluster is available once again.

Fixes #83838

droberts195 committed:
[ML] Retry datafeed searches on skipped CCS clusters

If a remote cluster has the skip_unavailable option set and is
unavailable during a datafeed search then we should retry rather
than risk using incomplete search results. Using incomplete search
results could lead to spurious anomalies being found that do not
make sense once the remote cluster is available once again.

Fixes #83838

elasticmachine commented:
Pinging @elastic/ml-core (Team:ML)

droberts195 committed:
Update docs/changelog/84052.yaml

dnhatn commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/scroll/ScrollDataExtractor.java:
```diff
@@ -214,12 +221,14 @@ void markScrollAsErrored() {
             client,
             () -> new SearchScrollRequestBuilder(client, SearchScrollAction.INSTANCE).setScroll(SCROLL_TIMEOUT).setScrollId(scrollId).get()
         );
+        checkForSkippedClusters(searchResponse);
```
We don't need to check this for subsequent scroll requests, but it's okay to leave this check as is here.

dnhatn reviewed (approved):
LGTM.

droberts195 committed:
Adding a test

droberts195 committed:
Merge branch 'master' into retry_datafeed_search_on_skipped_ccs_clusters

droberts195 commented on x-pack/plugin/ml/src/internalClusterTest/java/org/elasticsearch/xpack/ml/integration/DatafeedCcsIT.java:
```diff
@@ -0,0 +1,214 @@
+        clearSkipUnavailable();
+    }
+
+    public void testDatafeedWithCcsRemoteUnavailable() throws Exception {
```
This test often passes, and proves the code changes work.

However, the PR cannot be merged as-is because sometimes this test fails as follows:

```
	at __randomizedtesting.SeedInfo.seed([41E6F5EF31368549:15BE9EBA233BDCEA]:0)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
	at org.junit.Assert.assertThat(Assert.java:956)
	at org.junit.Assert.assertThat(Assert.java:923)
	at org.elasticsearch.test.InternalTestCluster.lambda$assertSearchContextsReleased$44(InternalTestCluster.java:2529)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1068)
	at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1041)
	at org.elasticsearch.test.InternalTestCluster.assertSearchContextsReleased(InternalTestCluster.java:2528)
	at org.elasticsearch.test.InternalTestCluster.assertAfterTest(InternalTestCluster.java:2470)
	at org.elasticsearch.test.AbstractMultiClustersTestCase.assertAfterTest(AbstractMultiClustersTestCase.java:139)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
```

The post-test assertions sometimes find unreleased search contexts.  So there must be somewhere in the datafeed code that doesn't release search contexts if a search fails.

In addition there are a couple of outstanding TODOs in this test, but I thought I'd push the code in its current form for visibility.

dnhatn replied:
I think the issue is that we need to clear scroll if the search response contains skipped clusters.

droberts195 committed:
Remove test for null list elements

droberts195 committed:
Cancel scrolls on incomplete results from remote clusters

Scrolls that fail due to partial results on local clusters will
be automatically cancelled. But where we choose to abort remote
searches that skipped data we need to explicitly cancel the
scrolls.

droberts195 committed:
Merge branch 'master' into retry_datafeed_search_on_skipped_ccs_clusters

droberts195 committed:
Fix TODOs

dimitris-athanasiou reviewed (approved):
LGTM
